### PR TITLE
[Merged by Bors] - Fix hare threshold calculations

### LINKED ISSUE #4499: Incorrect threshold calculation for Hare BFT consensus protocol
## Description
As part of our consensus logic there is a rule for calculating whether values should be considered. The original rule was that we need to see` >= f+1` support for a value, the intent of this rule is to ensure that at least one honest party supported a value. It was modified (see [here](https://community.spacemesh.io/t/hare-equivocations-defense/329/11) for discussion) to be ` >= f+1-k` where k is a count of known non-honest nodes. This modification ensures that we still only need support from a single honest node, even if we detect up to f dishonest nodes.

This rule was not being enforced correctly since we introduced a [CountInfo](https://github.com/spacemeshos/go-spacemesh/blob/2a7f0e0bbe517b7c1401e39892452592b12ce7a3/hare/algorithm.go#L142-L145) object which provided 2 different mechanisms to mark dishonest nodes, but when performing the threshold calculation it only considered dishonest nodes marked through one of the mechanisms, `dhCount` is not included in the `Meet` calculation.

```go
type CountInfo struct {
	// number of nodes that are honest, dishonest and known equivocators.
	// known equivocators are those the node only has eligibility proof but no hare message.
	numHonest, numDishonest, numKE int
	hCount, dhCount, keCount       int
}

func (ci *CountInfo) Meet(threshold int) bool {
	// at least one honest party with good message
	return ci.hCount > 0 && ci.hCount+ci.keCount >= threshold
}
```

This logic is used in commit, pre-round, status and notify trackers, so they are likely all not functioning correctly.

### PR DESCRIPTION
**Pull request by piersy**
## Motivation
Closes #4499 

## Changes
Modifies the Meet method calculation for CountInfo

## Test Plan
Added tests for all affected trackers, and for the CountInfo object

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Review by dshulyak — DISMISSED** [review]

File: hare/algorithm_test.go [hunk 1]
```diff
@@ -174,6 +175,44 @@ func (mev *mockEligibilityValidator) ValidateEligibilityGossip(context.Context,
 	return atomic.LoadInt32(&mev.valid) != 0
 }
 
+// We assert the the theshold X is met a long as there is one honest vote and
+// the sum of honest, dishonest and known equivocator is >= X.
```
Comment (countvonzero) [inline hare/algorithm_test.go]: this description is not exactly right. the condition also requires at least 1 honest vote.

File: hare/algorithm_test.go [hunk 2]
```diff
@@ -174,6 +175,44 @@ func (mev *mockEligibilityValidator) ValidateEligibilityGossip(context.Context,
...
+func TestCountInfoMeetsTrheshold(t *testing.T) {
```
Comment (countvonzero) [inline hare/algorithm_test.go:183]: i think the test below is sufficient to deliver your point.
```suggestion
func TestCountInfoMeetsTrheshold(t *testing.T) {
        c := &CountInfo{}
	c.IncKnownEquivocator(1)
	require.False(t, c.Meet(3))
	c.IncDishonest(1)
	require.False(t, c.Meet(3))
	c.IncHonest(1)
	require.True(t, c.Meet(3))
}
```

**Review by countvonzero — APPROVED** [review]

File: hare/algorithm_test.go (hunk 1 above)
Comment (piersy) [reply hare/algorithm_test.go]: That's in there right?
>We assert the the theshold X is met **a long as there is one honest vote** and

Although I have noticed a typo **a -> as**

Comment (piersy) [reply hare/algorithm_test.go]: Oh do you mean it should be *at least one* ?

File: hare/algorithm_test.go (hunk 2 above)
Comment (piersy) [reply hare/algorithm_test.go:183]: This test doesn't assert that at least one honest vote is required to meet the threshold.

I.E the following implementation of Meet would pass
```
ci.hCount+ci.dhCount+ci.keCount >= threshold
```

**Comment by piersy** [comment]
bors cancel

**Comment by piersy** [comment]
bors merge

**Comment by dshulyak** [comment]
bors try

**Comment by piersy** [comment]
bors merge

**Comment by piersy** [comment]
bors merge

**Pull request closed without merging**